spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java

```diff
@@ -0,0 +1,207 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark;
+
+import java.io.Serializable;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import org.apache.iceberg.ChangelogOperation;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterators;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.RowFactory;
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema;
+
+/**
+ * An iterator that transforms rows from changelog tables within a single Spark task. It assumes
+ * that rows are sorted by identifier columns and change type.
+ *
+ * <p>It removes the carry-over rows. Carry-over rows are the result of a removal and insertion of
+ * the same row within an operation because of the copy-on-write mechanism. For example, given a
+ * file which contains row1 (id=1, data='a') and row2 (id=2, data='b'). A copy-on-write delete of
+ * row2 would require erasing this file and preserving row1 in a new file. The change-log table
+ * would report this as (id=1, data='a', op='DELETE') and (id=1, data='a', op='INSERT'), despite it
+ * not being an actual change to the table. The iterator finds the carry-over rows and removes them
+ * from the result.
+ *
+ * <p>This iterator also finds delete/insert rows which represent an update, and converts them into
+ * update records. For example, these two rows
+ *
+ * <ul>
+ *   <li>(id=1, data='a', op='DELETE')
+ *   <li>(id=1, data='b', op='INSERT')
+ * </ul>
+ *
+ * <p>will be marked as update-rows:
+ *
+ * <ul>
+ *   <li>(id=1, data='a', op='UPDATE_BEFORE')
+ *   <li>(id=1, data='b', op='UPDATE_AFTER')
+ * </ul>
+ */
+public class ChangelogIterator implements Iterator<Row>, Serializable {
+  private static final String DELETE = ChangelogOperation.DELETE.name();
+  private static final String INSERT = ChangelogOperation.INSERT.name();
+  private static final String UPDATE_BEFORE = ChangelogOperation.UPDATE_BEFORE.name();
+  private static final String UPDATE_AFTER = ChangelogOperation.UPDATE_AFTER.name();
+
+  private final Iterator<Row> rowIterator;
+  private final int changeTypeIndex;
+  private final List<Integer> identifierFieldIdx;
+
+  private Row cachedRow = null;
+  private int[] indicesForIdentifySameRow = null;
+
+  private ChangelogIterator(
+      Iterator<Row> rowIterator, int changeTypeIndex, List<Integer> identifierFieldIdx) {
```

Comment:
Optional: I'd probably pass a schema into this method and make it responsible for computing indices rather than relying on input. This would allow us to compute things like `indicesForIdentifySameRow` right away instead of having an if condition for every row. It is optional but it seems safer and would make this class more independent. The downside is the need to compute the indices on each creation but I don't think it matters cause it is done once per Spark partition.